File: 이지수/Python/Baekjoon_1025.py
# 1025
import sys

input = sys.stdin.readline
n, m = map(int, input().split())

def getMap(n) :
    inputs = []
    for _ in range(n):
        ip = []
        [ip.append(int(num)) for num in input().rstrip()]
        inputs.append(ip)
    return inputs

def getBiggest(maps, rd, cd) :
    results = [-1]
    for i in range(n) :
        for j in range(m) :
            results.append(getResult(maps, i, j, rd, cd))
    return max(results)

def getResult(maps, tx, ty, rd, cd) :
    results = []
    types = [(-1, -1), (-1, 1), (1, -1), (1, 1)]
    if cd == 0 and rd == 0 :
        return maps[tx][ty] if maps[tx][ty] ** 0.5 % 1 == 0 else -1
    for t in types :
        result = []
        rs = 0
        if rd == 0 :
            i = tx
            for j in range(ty, m if t[1] == 1 else -1, cd * t[1]) :
                rs *= 10
                rs += maps[i][j]
                result.append(rs)
        elif cd == 0 :
            j = ty
            for i in range(tx, n if t[0] == 1 else -1, rd * t[0]) :
                rs *= 10
                rs += maps[i][j]
                result.append(rs)
        else :
            i, j = tx, ty
            while 0 <= i < n and 0 <= j < m :
                rs *= 10
                rs += maps[i][j]
                i += rd * t[0]; j += cd * t[1]
                result.append(rs)
        for rs in result :
            results.append(rs if rs ** 0.5 % 1 == 0 else -1)    
    return max(results)

def solution(n, m) :
    maps = getMap(n)
    biggest = -1
    for i in range(n + 1) :
        for j in range(m + 1) :
            result = getBiggest(maps, i, j) 
            biggest = result if biggest < result else biggest
    return biggest

print(solution(n, m))

File: 이지수/Swift/Baekjoon_1025.swift
// 1025
import Foundation

@inline(__always)
func getMap(_ n: Int) -> [[Int]] {
    var map: [[Int]] = []
    for _ in 0..<n {
        map.append(readLine()!.reduce([]) { $0 + [Int(String($1))!] })
    }
    return map
}

@inline(__always)
func getMaxIn(cDist: Int, rDist: Int, _ size: (n: Int, m:Int), _ map: [[Int]]) -> Int {
    var maxValue = -1
    for i in 0..<size.n {
        for j in 0..<size.m {
            maxValue = max(getMaxAt(point: (i, j), dist: (cDist, rDist), size, map), maxValue)
        }
    }
    return maxValue
}

@inline(__always)
func getMaxAt(point target: (x: Int, y: Int), dist: (c: Int, r: Int), _ size: (n: Int, m: Int), _ map: [[Int]]) -> Int {
    guard dist.c != 0 || dist.r != 0 else {
        return sqrt(Double(map[target.x][target.y])).truncatingRemainder(dividingBy: 1) == 0 ? map[target.x][target.y] : -1
    }
    var results: [Int] = []
    let types: [(Int, Int)] = dist.c * dist.r == 0 ? [(-1, -1), (1, 1)] : [(1, 1), (1, -1), (-1, -1), (-1, 1)]
    for type in types {
        var result: [Int] = []
        var current = 0
        if dist.c == 0 {
            let i = target.x
            Array(stride(from: target.y, to: type.1 > 0 ? size.m : -1, by: dist.r * type.1)).forEach { j in
                current *= 10
                current += map[i][j]
                result.append(current)
            }
        }
        else if dist.r == 0 {
            let j = target.y
            Array(stride(from: target.x, to: type.0 > 0 ? size.n : -1, by: dist.c * type.0)).forEach { i in
                current *= 10
                current += map[i][j]
                result.append(current)
            }
        }
        else {
            var (i, j) = target
            while (0..<size.n ~= i && 0..<size.m ~= j) {
                current *= 10
                current += map[i][j]
                result.append(current)
                i += (dist.c * type.0); j += (dist.r * type.1)
            }
        }
        results += result.filter() { sqrt(Double($0)).truncatingRemainder(dividingBy: 1) == 0 }
    }
    return results.max() ?? -1
}

@inline(__always)
func solution(n: Int, m: Int) -> Int {
    let map = getMap(n)
    var maxValue: Int = -1
    Array(0..<n).forEach { i in
        Array(0..<m).forEach { j in
            maxValue = max(maxValue, getMaxIn(cDist: i, rDist: j, (n, m), map))
        }
    }
    return maxValue
}

let inputs = readLine()!.components(separatedBy: " ").map() { Int($0)! }
print(solution(n: inputs[0], m: inputs[1]))
